### PR TITLE
feat(select): add isClearable

### LINKED ISSUE #2239: [Feature Request] Add isClearable for Select
### Is your feature request related to a problem? Please describe.

This option is available in the new Autocomplete but not in the Select  

### Describe the solution you'd like

```
 <Select
      isClearable
      label="Favorite Animal"
      placeholder="Select an animal"
     ...
    >
```

### Describe alternatives you've considered

Currently the only way to do this is using `endContent` with a button

### Screenshots or Videos

_No response_

### PR DESCRIPTION
**Pull request by abhinav700**
Closes #2239 

## 📝 Description
Added clear button functionality to select component.
Screenshot from 2024-09-13 18-02-21
Screenshot from 2024-09-13 18-02-11
Screenshot from 2024-09-13 18-01-17

## Summary by CodeRabbit

- **New Features**
	- Introduced a selection component for choosing favorite animals with a clearable feature.
	- Added a `PetBoldIcon` component to enhance the dropdown menu.
	- Users can now reset their selection easily with the clearable functionality.
	- Documentation updated to include the new `isClearable` property and `onClear` event handler for the Select component.

- **Bug Fixes**
	- Enhanced the styling and responsiveness of the clear button across different component sizes.

### CODE REVIEW

**Comment by abhinav700** [comment]
@wingkwong for the tests that are failing, I don't think they are related to the changes that I have made.

When I tried to run the tests on the canary branch( which is in sync and I have pulled those changes locally), the same tests are failing there also even though I have not made any changes to it.

File: packages/components/select/src/use-select.ts
Comment (wingkwong) [inline packages/components/select/src/use-select.ts:668]: image

in 2.5.0, clear button won't be focusable.

File: packages/core/theme/src/components/select.ts
Comment (wingkwong) [inline packages/core/theme/src/components/select.ts:172]: don't use `rtl:*`. there use `start` and `end` for RTL. 

see https://tailwindcss.com/blog/tailwindcss-v3-3#simplified-rtl-support-with-logical-properties

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/core/theme/src/components/select.ts
Comment (abhinav700) [reply packages/core/theme/src/components/select.ts:172]: okay I will look into that.

File: packages/components/select/src/select.tsx
Comment (abhinav700) [reply packages/components/select/src/select.tsx:141]: I can change the code. Can you specify if we should use `endConten`t as clear button or not?

File: packages/components/select/src/use-select.ts
Comment (abhinav700) [reply packages/components/select/src/use-select.ts:668]: @wingkwong so should I remove the `focusable` property from clear button?

Comment (wingkwong) [reply packages/components/select/src/use-select.ts:668]: check the latest input implementation. `tabIndex` should be `-1`.

File: packages/components/select/src/select.tsx
Comment (wingkwong) [reply packages/components/select/src/select.tsx]: ideally we should support both at the same time

Comment (abhinav700) [reply packages/components/select/src/select.tsx]: I implement that by creating a `clearButtonIcon` prop.

in that case, clear button will be visible if `isClearable`, `clearButtonIcon` or `onClear` is provided.

Comment (wingkwong) [reply packages/components/select/src/select.tsx]: Here's what I thought 

- Clear button should be shown when `isClearable` is set and with selected value.
- `onClear` is not necessarily provided by users. It's just an action to execute after clearing. It's optional and it doesn't affect the visibility of clear button.
- I don't think we need `clearButtonIcon` at this moment.

Comment (wingkwong) [reply packages/components/select/src/select.tsx]: you may check PR2305 as a reference.

Comment (wingkwong) [reply packages/components/select/src/select.tsx]: sure

Comment (abhinav700) [reply packages/components/select/src/select.tsx]: okay

Comment (abhinav700) [reply packages/components/select/src/select.tsx]: Is the alignment okay?

LTR:
Screenshot from 2024-10-01 10-36-04

RTL:
image

Comment (abhinav700) [reply packages/components/select/src/select.tsx]: I can also modify it to display `endContent` before clear button and then align the` endContent` with the text and clear button with the selector icon.

I am not sure which one to implement though, so would need a confirmation from your side first.

Comment (wingkwong) [reply packages/components/select/src/select.tsx]: sorry for late reply. it looks good. but i think you don't need to include two examples. just one is enough, i.e. with endcontent

Comment (abhinav700) [reply packages/components/select/src/select.tsx]: okay I will correct that.

File: packages/components/select/stories/select.stories.tsx
Comment (wingkwong) [reply packages/components/select/stories/select.stories.tsx:1017]: @abhinav700 please add that also

File: packages/components/select/src/select.tsx
Comment (wingkwong) [inline packages/components/select/src/select.tsx:32]: wt's that for?

File: apps/docs/content/components/select/isClearable.ts
Comment (wingkwong) [inline apps/docs/content/components/select/isClearable.ts:1]: rename to `is-clearable.ts`

File: .changeset/olive-buckets-own.md
Comment (wingkwong) [inline .changeset/olive-buckets-own.md:5]: change to

```
add `isClearable` and `onClear` prop to Select component (#2239)
```

File: .changeset/olive-buckets-own.md
Comment (wingkwong) [inline .changeset/olive-buckets-own.md:2]: include theme package as well

**Review by wingkwong — CHANGES_REQUESTED** [review]
since clearButton in select theme is introduced, please also update the min theme version in peerDependencies in select package.json.

File: packages/components/select/src/select.tsx
Comment (abhinav700) [reply packages/components/select/src/select.tsx:32]: idk, maybe it appeared when I updated the branch. will check again

File: packages/components/select/stories/select.stories.tsx
Comment (abhinav700) [reply packages/components/select/stories/select.stories.tsx:1017]: I get linting errors while comitting this change.
image

File: apps/docs/content/docs/components/select.mdx
Comment (wingkwong) [inline apps/docs/content/docs/components/select.mdx:153]:
```diff
-it will have a clear button which will be visible only when some value is selected.
+it will have a clear button which will be visible only when a value is selected.
```

File: packages/components/select/package.json
Comment (wingkwong) [inline packages/components/select/package.json:38]: currently it's `2.2.11`. I think it should be `2.3.0` since it is a minor bump.

File: packages/components/select/stories/select.stories.tsx
Comment (wingkwong) [inline packages/components/select/stories/select.stories.tsx:1016]: please add `// eslint-disable-next-line no-console` above this line.

File: packages/components/select/src/use-select.ts
Comment (wingkwong) [inline packages/components/select/src/use-select.ts:83]: i think this statement is no longer correct. now the endContent is placed next to clear button. these 3 lines can be deleted.

**Review by wingkwong — CHANGES_REQUESTED** [review]
clear button shouldn't be focused. you may try the following steps.

1. open a select, and choose a value.
2. click somewhere else to blur, then press tab. select is focused, close button is not focused (which is expected)
3. click the value by clicking clear button
4. open a select, and choose a value
5. click somewhere else to blur, then press tab. select is focused, close button is focused (which is not expected) 

image

**Comment by abhinav700** [comment]
> 5. click somewhere else to blur, then press tab. select is focused, close button is focused (which is not expected)

Given that tabIndex is set to -1, that shouldn't happen. WIll look into it.

**Comment by abhinav700** [comment]
> clear button shouldn't be focused.
@wingkwong I have fixed this issue
can you check again?

**Review by wingkwong — APPROVED** [review]

File: packages/core/theme/src/components/select.ts
Comment (ryo-manba) [inline packages/core/theme/src/components/select.ts:38]: Could you align the [input's clear button](https://nextui.org/docs/components/input#clear-button)?
It looks like there's no hover style applied.

File: apps/docs/content/components/select/is-clearable.ts
Comment (ryo-manba) [inline apps/docs/content/components/select/is-clearable.ts:62]: I think it’s clearer not to specify endContent, and it’s more common as an example.

File: apps/docs/content/components/select/is-clearable.ts
Comment (ryo-manba) [inline apps/docs/content/components/select/is-clearable.ts:63]: `console.log` is unnecessary.

**Review by ryo-manba — CHANGES_REQUESTED** [review]

File: apps/docs/content/components/select/is-clearable.ts
Comment (abhinav700) [reply apps/docs/content/components/select/is-clearable.ts:63]: Should I remove the entire line or just leave the `onClear` as an empty function?

**Comment by wingkwong** [comment]
mark it on hold until the doc structure revamp pr got merged to beta branch

**Comment by abhinav700** [comment]
ok

File: apps/docs/content/components/select/is-clearable.ts
Comment (ryo-manba) [reply apps/docs/content/components/select/is-clearable.ts:62]: I don't think it's necessary for Storybook either.

File: apps/docs/content/components/select/is-clearable.ts
Comment (ryo-manba) [reply apps/docs/content/components/select/is-clearable.ts:63]: I think it's fine to delete the entire row.

File: apps/docs/content/components/select/is-clearable.ts
Comment (abhinav700) [reply apps/docs/content/components/select/is-clearable.ts:62]: okay

File: packages/core/theme/src/components/select.ts
Comment (wingkwong) [reply packages/core/theme/src/components/select.ts:38]: @abhinav700 is it handled?

Comment (abhinav700) [reply packages/core/theme/src/components/select.ts:38]: Yup, hover style is now fixed for select

Regarding input component, isn't it already aligned correctly

image

**Pull request closed without merging**
